src/sock.c

```diff
@@ -282,15 +282,14 @@ static void read_conn(struct mg_connection *c) {
       n = recv_raw(c, (char *) &c->rtls.buf[c->rtls.len],
                    c->rtls.size - c->rtls.len);
       // MG_DEBUG(("%lu %ld", c->id, n));
-      if (n == MG_IO_ERR) {
+      if (n == MG_IO_ERR && mg_tls_pending(c) == 0 && c->rtls.len == 0) {
+        // Close only if we have fully drained both raw (rtls) and TLS buffers
         c->is_closing = 1;
-      } else if (n > 0) {
-        c->rtls.len += (size_t) n;
+      } else {
+        if (n > 0) c->rtls.len += (size_t) n;
         if (c->is_tls_hs) mg_tls_handshake(c);
         if (c->is_tls_hs) return;
         n = mg_tls_recv(c, buf, len);
-      } else if (n == MG_IO_WAIT) {
-        n = mg_tls_recv(c, buf, len);
       }
```

Comment:
mg_tls_handshake() used to execute on `n > 0`, now that path is taken as long as there is no IO_ERR; even on IO_WAIT, overriding the call to mg_tls_recv().
Looks cleaner, as long as that is what's intended.